File: src/api_medic/core/render/markdown.py
"""Markdown renderer for Reports.

GitHub-flavored markdown. Designed to render cleanly when pasted into a
GitHub issue, a Slack message, or an email body. Severity is conveyed by
[CRITICAL]/[WARNING]/[INFO] tags rather than emoji.
"""

from __future__ import annotations

from typing import Any

from ..models import Finding, Report, Severity, TimingBreakdown

_SEVERITY_TAG: dict[Severity, str] = {
    "critical": "[CRITICAL]",
    "warning": "[WARNING]",
    "info": "[INFO]",
}


def render_markdown(report: Report) -> str:
    parts: list[str] = []
    parts.append("# api-medic — diagnostic report")
    parts.append("")
    parts.append(f"`{report.request.method} {report.request.url}` → " + _status_str(report))
    parts.append("")
    parts.append(_metrics_table(report))
    parts.append("")
    timing = _timing_table(report.timing)
    if timing:
        parts.append("## Timing")
        parts.append("")
        parts.append(timing)
        parts.append("")
    parts.append("## Findings")
    parts.append("")
    if not report.findings:
        parts.append("_No findings — the request looks healthy._")
    else:
        for f in report.findings:
            parts.append(_finding_block(f))
            parts.append("")
    return "\n".join(parts).rstrip() + "\n"


def _status_str(report: Report) -> str:
    if report.response is None:
        return "_no response_"
    return f"`{report.response.status_code} {report.response.status_text}`"


def _metrics_table(report: Report) -> str:
    latency = _fmt_latency(report.timing.total_ms)
    body = _fmt_bytes(report.response.body_size_bytes) if report.response else "—"
    protocol = report.response.protocol if report.response else "—"
    findings = _fmt_findings_count(report.findings)
    return (
        "| Latency | Body | Protocol | Findings |\n"
        "|---------|------|----------|----------|\n"
        f"| {latency} | {body} | {protocol} | {findings} |"
    )


def _timing_table(t: TimingBreakdown) -> str:
    rows: list[tuple[str, float]] = []
    for label, val in (
        ("DNS", t.dns_ms),
        ("Connect", t.connect_ms),
        ("TLS", t.tls_ms),
        ("TTFB", t.ttfb_ms),
        ("Download", t.download_ms),
    ):
        if val is not None:
            rows.append((label, val))
    if not rows and t.total_ms is None:
        return ""
    out = ["| Phase | Duration |", "|-------|---------:|"]
    for label, val in rows:
        out.append(f"| {label} | {val:.0f} ms |")
    if t.total_ms is not None:
        out.append(f"| **Total** | **{t.total_ms:.0f} ms** |")
    return "\n".join(out)


def _finding_block(f: Finding) -> str:
    lines = [f"### {_SEVERITY_TAG[f.severity]} {f.title}", f"**`{f.id}`**", "", f.explanation]
    if f.evidence:
        lines.append("")
        lines.append("**Evidence:**")
        for k, v in f.evidence.items():
            lines.append(f"- `{k}`: `{_evidence_value(v)}`")
    if f.suggested_fix:
        lines.append("")
        lines.append(f"**Suggested fix:** {f.suggested_fix}")
    return "\n".join(lines)


def _evidence_value(v: Any) -> str:
    if isinstance(v, str):
        return v
    if isinstance(v, (int, float, bool)):
        return str(v)
    if v is None:
        return "null"
    import json as _json

    return _json.dumps(v)


def _fmt_latency(total: float | None) -> str:
    if total is None:
        return "—"
    if total < 1000:
        return f"{total:.0f} ms"
    return f"{total / 1000:.2f} s"


def _fmt_bytes(n: int) -> str:
    if n < 1024:
        return f"{n} B"
    if n < 1024 * 1024:
        return f"{n / 1024:.1f} kB"
    return f"{n / (1024 * 1024):.1f} MB"


def _fmt_findings_count(findings: list[Finding]) -> str:
    if not findings:
        return "0"
    by_severity: dict[Severity, int] = {"critical": 0, "warning": 0, "info": 0}
    for f in findings:
        by_severity[f.severity] += 1
    if by_severity["critical"]:
        return f"{by_severity['critical']} critical"
    if by_severity["warning"]:
        return f"{by_severity['warning']} warning"
    return f"{by_severity['info']} info"

File: frontend/src/lib/markdown.ts
// Markdown renderer for Reports. Mirrors src/api_medic/core/render/markdown.py
// so Export-markdown produces equivalent output across CLI and web/extension
// surfaces. Severity is conveyed by [CRITICAL]/[WARNING]/[INFO] tags rather
// than emoji to render cleanly in GitHub issues, Slack, or email.

import type { Finding, Report, Severity, TimingBreakdown } from "./types";

const SEVERITY_TAG: Record<Severity, string> = {
  critical: "[CRITICAL]",
  warning: "[WARNING]",
  info: "[INFO]",
};

export function renderMarkdown(report: Report): string {
  const parts: string[] = [];
  parts.push("# api-medic — diagnostic report");
  parts.push("");
  parts.push(
    `\`${report.request.method} ${report.request.url}\` → ${statusStr(report)}`,
  );
  parts.push("");
  parts.push(metricsTable(report));
  parts.push("");
  const timing = timingTable(report.timing);
  if (timing) {
    parts.push("## Timing");
    parts.push("");
    parts.push(timing);
    parts.push("");
  }
  parts.push("## Findings");
  parts.push("");
  const findings = report.findings ?? [];
  if (findings.length === 0) {
    parts.push("_No findings — the request looks healthy._");
  } else {
    for (const f of findings) {
      parts.push(findingBlock(f));
      parts.push("");
    }
  }
  return parts.join("\n").replace(/\s+$/, "") + "\n";
}

function statusStr(report: Report): string {
  if (!report.response) return "_no response_";
  return `\`${report.response.status_code} ${report.response.status_text}\``;
}

function metricsTable(report: Report): string {
  const latency = fmtLatency(report.timing.total_ms ?? null);
  const body = report.response
    ? fmtBytes(report.response.body_size_bytes)
    : "—";
  const protocol = report.response ? report.response.protocol : "—";
  const findings = fmtFindingsCount(report.findings ?? []);
  return (
    "| Latency | Body | Protocol | Findings |\n" +
    "|---------|------|----------|----------|\n" +
    `| ${latency} | ${body} | ${protocol} | ${findings} |`
  );
}

function timingTable(t: TimingBreakdown): string {
  const rows: Array<[string, number]> = [];
  const phases: Array<[string, number | null | undefined]> = [
    ["DNS", t.dns_ms],
    ["Connect", t.connect_ms],
    ["TLS", t.tls_ms],
    ["TTFB", t.ttfb_ms],
    ["Download", t.download_ms],
  ];
  for (const [label, val] of phases) {
    if (val !== null && val !== undefined) rows.push([label, val]);
  }
  if (rows.length === 0 && (t.total_ms === null || t.total_ms === undefined)) {
    return "";
  }
  const out = ["| Phase | Duration |", "|-------|---------:|"];
  for (const [label, val] of rows) {
    out.push(`| ${label} | ${val.toFixed(0)} ms |`);
  }
  if (t.total_ms !== null && t.total_ms !== undefined) {
    out.push(`| **Total** | **${t.total_ms.toFixed(0)} ms** |`);
  }
  return out.join("\n");
}

function findingBlock(f: Finding): string {
  const lines: string[] = [
    `### ${SEVERITY_TAG[f.severity]} ${f.title}`,
    `**\`${f.id}\`**`,
    "",
    f.explanation,
  ];
  if (f.evidence && Object.keys(f.evidence).length > 0) {
    lines.push("");
    lines.push("**Evidence:**");
    for (const [k, v] of Object.entries(f.evidence)) {
      lines.push(`- \`${k}\`: \`${evidenceValue(v)}\``);
    }
  }
  if (f.suggested_fix) {
    lines.push("");
    lines.push(`**Suggested fix:** ${f.suggested_fix}`);
  }
  return lines.join("\n");
}

function evidenceValue(v: unknown): string {
  if (typeof v === "string") return v;
  if (typeof v === "number" || typeof v === "boolean") return String(v);
  if (v === null || v === undefined) return "null";
  return JSON.stringify(v);
}

function fmtLatency(total: number | null): string {
  if (total === null || total === undefined) return "—";
  if (total < 1000) return `${total.toFixed(0)} ms`;
  return `${(total / 1000).toFixed(2)} s`;
}

function fmtBytes(n: number): string {
  if (n < 1024) return `${n} B`;
  if (n < 1024 * 1024) return `${(n / 1024).toFixed(1)} kB`;
  return `${(n / (1024 * 1024)).toFixed(1)} MB`;
}

function fmtFindingsCount(findings: Finding[]): string {
  if (findings.length === 0) return "0";
  const counts: Record<Severity, number> = {
    critical: 0,
    warning: 0,
    info: 0,
  };
  for (const f of findings) counts[f.severity] += 1;
  if (counts.critical) return `${counts.critical} critical`;
  if (counts.warning) return `${counts.warning} warning`;
  return `${counts.info} info`;
}
